File: benchmark/ts/random_bench/distributions.ts
/**
 * Random number generation benchmarks: uniform, normal, randint
 */

import { BenchmarkRunner } from '../lib/BenchmarkRunner.js';
import { BENCHMARK_CONFIG, ARRAY_SIZES, RANDOM_SEED } from '../lib/config.js';
import { formatSize } from '../lib/utils.js';
import type { OperationResults, BenchmarkDataPoint } from '../lib/types.js';

// @ts-expect-error - Types are in dist/index.d.ts but TS can't resolve mjs
import { loadWasmModule } from '../../../numjs.mjs';
// @ts-expect-error
import * as random from '../../../numjs.mjs';

async function benchmarkUniform(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of ARRAY_SIZES) {
    process.stdout.write(`    uniform size=${formatSize(size)}... `);

    // Seed the generator before each benchmark for reproducibility
    random.seed(RANDOM_SEED);

    const result = runner.runSync(
      () => null,
      () => {
        const arr = random.uniform(0, 1, [size]);
        const val = arr.sum();
        arr.dispose();
        return val;
      }
    );

    results.push({
      params: { size },
      meanMs: result.meanMs,
      minMs: result.minMs,
      maxMs: result.maxMs,
      stdMs: result.stdMs,
      iterations: result.iterations,
      result: result.result,
    });

    console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
  }

  return { id: 'uniform', name: 'Uniform', results };
}

async function benchmarkNormal(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of ARRAY_SIZES) {
    process.stdout.write(`    normal size=${formatSize(size)}... `);

    random.seed(RANDOM_SEED);

    const result = runner.runSync(
      () => null,
      () => {
        const arr = random.normal(0, 1, [size]);
        const val = arr.sum();
        arr.dispose();
        return val;
      }
    );

    results.push({
      params: { size },
      meanMs: result.meanMs,
      minMs: result.minMs,
      maxMs: result.maxMs,
      stdMs: result.stdMs,
      iterations: result.iterations,
      result: result.result,
    });

    console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
  }

  return { id: 'normal', name: 'Normal', results };
}

async function benchmarkRandint(runner: BenchmarkRunner): Promise<OperationResults> {
  const results: BenchmarkDataPoint[] = [];

  for (const size of ARRAY_SIZES) {
    process.stdout.write(`    randint size=${formatSize(size)}... `);

    random.seed(RANDOM_SEED);

    const result = runner.runSync(
      () => null,
      () => {
        const arr = random.randint(0, 100, [size]);
        const val = arr.sum();
        arr.dispose();
        return val;
      }
    );

    results.push({
      params: { size },
      meanMs: result.meanMs,
      minMs: result.minMs,
      maxMs: result.maxMs,
      stdMs: result.stdMs,
      iterations: result.iterations,
      result: result.result,
    });

    console.log(`${result.meanMs.toFixed(4)} ms (${result.iterations} iters)`);
  }

  return { id: 'randint', name: 'Random Integer', results };
}

export async function runRandomBenchmarks(): Promise<OperationResults[]> {
  console.log('  Loading WASM module...');
  await loadWasmModule();

  const runner = new BenchmarkRunner(BENCHMARK_CONFIG);
  const results: OperationResults[] = [];

  console.log('  Benchmarking uniform...');
  results.push(await benchmarkUniform(runner));

  console.log('  Benchmarking normal...');
  results.push(await benchmarkNormal(runner));

  console.log('  Benchmarking randint...');
  results.push(await benchmarkRandint(runner));

  return results;
}

File: benchmark/python/random_bench/distributions.py
"""
Random number generation benchmarks: uniform, normal, randint
"""

import numpy as np
from ..lib.benchmark_runner import BenchmarkRunner
from ..lib.config import BENCHMARK_CONFIG, ARRAY_SIZES, RANDOM_SEED


def format_size(size: int) -> str:
    """Format a size number for display."""
    if size >= 1_000_000:
        return f"{size // 1_000_000}M"
    elif size >= 1_000:
        return f"{size // 1_000}K"
    return str(size)


def benchmark_uniform(runner: BenchmarkRunner) -> dict:
    """Benchmark np.random.uniform."""
    results = []

    for size in ARRAY_SIZES:
        print(f"    uniform size={format_size(size)}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        result = runner.run(lambda s=size: np.random.uniform(0, 1, s))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'uniform', 'name': 'Uniform', 'results': results}


def benchmark_normal(runner: BenchmarkRunner) -> dict:
    """Benchmark np.random.normal."""
    results = []

    for size in ARRAY_SIZES:
        print(f"    normal size={format_size(size)}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        result = runner.run(lambda s=size: np.random.normal(0, 1, s))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'normal', 'name': 'Normal', 'results': results}


def benchmark_randint(runner: BenchmarkRunner) -> dict:
    """Benchmark np.random.randint."""
    results = []

    for size in ARRAY_SIZES:
        print(f"    randint size={format_size(size)}... ", end='', flush=True)

        np.random.seed(RANDOM_SEED)
        result = runner.run(lambda s=size: np.random.randint(0, 100, s))

        results.append({
            'params': {'size': size},
            'meanMs': result.mean_ms,
            'minMs': result.min_ms,
            'maxMs': result.max_ms,
            'stdMs': result.std_ms,
            'iterations': result.iterations,
        })

        print(f"{result.mean_ms:.4f} ms ({result.iterations} iters)")

    return {'id': 'randint', 'name': 'Random Integer', 'results': results}


def run_random_benchmarks() -> list[dict]:
    """Run all random benchmarks."""
    runner = BenchmarkRunner(BENCHMARK_CONFIG)
    results = []

    print("  Benchmarking uniform...")
    results.append(benchmark_uniform(runner))

    print("  Benchmarking normal...")
    results.append(benchmark_normal(runner))

    print("  Benchmarking randint...")
    results.append(benchmark_randint(runner))

    return results


if __name__ == '__main__':
    import json
    results = run_random_benchmarks()
    print(json.dumps(results, indent=2))
